### DISCUSSION TITLE
About custom distance metric

### DISCUSSION THREAD

**Discussion by zineb198** [facebookresearch/faiss#1904] [Q&A]
Hello,

Would it be possible to use the faiss package with user-defined metrics?

Thanks a lot for this package :D

**Comment by GioiaZheng** [comment]
FAISS does not support arbitrary user-defined (callback) metrics — search runs in C++/SIMD, so there's no hook for a Python distance function. The built-in metrics are `METRIC_L2`, `METRIC_INNER_PRODUCT`, and a set of extra ones: `METRIC_L1`, `METRIC_Linf`, `METRIC_Lp` (with `metric_arg = p`), `METRIC_Canberra`, `METRIC_BrayCurtis`, `METRIC_JensenShannon`, plus Jaccard/Hamming for binary indexes.

Practical workarounds:
- **Reduce your metric to a built-in via preprocessing.** Many "custom" metrics are L2/IP after a transform: cosine → L2-normalize the vectors and use `METRIC_INNER_PRODUCT`; Mahalanobis → apply the Cholesky/whitening transform (e.g. a `LinearTransform`) then `METRIC_L2`.
- **Monotonic transforms preserve ranking.** If your metric is a monotonic function of L2 or IP, the neighbor order is unchanged — search with the built-in and recompute exact scores only if you need the values.
- **Coarse search + exact rerank.** Use a built-in proxy metric to fetch top-k candidates, then rerank that small set with your exact custom metric in NumPy. This is the general escape hatch when the metric can't be transformed.
